File: CSC310-Cpp/Module15-Computational-Geometry/debug02/convex_hull.cpp
// convex_hull.cpp
// Convex hull by Graham's scan (CLRS 33.3). The anchor is the lowest point
// (leftmost on ties); the remaining points are sorted by polar angle around
// the anchor and scanned with a stack, popping whenever the top of the stack
// would not make a left turn. The hull is reported counter-clockwise from
// the anchor and contains only its corner points.
#include <algorithm>
#include <iostream>
#include <string>
#include <vector>

struct Point {
    long x, y;
};

// Cross product (b - a) x (c - a): > 0 for a left turn a->b->c, 0 if collinear.
long cross(Point a, Point b, Point c) {
    return (b.x - a.x) * (c.y - a.y) - (b.y - a.y) * (c.x - a.x);
}

long distSquared(Point a, Point b) {
    return (a.x - b.x) * (a.x - b.x) + (a.y - b.y) * (a.y - b.y);
}

std::vector<Point> convexHull(std::vector<Point> pts) {
    if (pts.size() < 3) {
        return pts;
    }
    // Anchor: minimum y, then minimum x.
    auto anchorIt = std::min_element(pts.begin(), pts.end(), [](Point a, Point b) {
        return a.y < b.y || (a.y == b.y && a.x < b.x);
    });
    std::swap(*anchorIt, pts[0]);
    const Point anchor = pts[0];

    // Sort by polar angle around the anchor.
    std::sort(pts.begin() + 1, pts.end(), [&](Point a, Point b) {
        long c = cross(anchor, a, b);
        if (c != 0) return c > 0;
        return distSquared(anchor, a) > distSquared(anchor, b);
    });

    std::vector<Point> hull;
    for (const Point& p : pts) {
        while (hull.size() >= 2 && cross(hull[hull.size() - 2], hull.back(), p) <= 0) {
            hull.pop_back();
        }
        hull.push_back(p);
    }
    return hull;
}

// Twice the signed area (shoelace formula); positive for counter-clockwise.
long twiceArea(const std::vector<Point>& poly) {
    long sum = 0;
    for (std::size_t i = 0; i < poly.size(); ++i) {
        const Point& a = poly[i];
        const Point& b = poly[(i + 1) % poly.size()];
        sum += a.x * b.y - b.x * a.y;
    }
    return sum;
}

void report(const std::string& label, const std::vector<Point>& pts) {
    std::vector<Point> hull = convexHull(pts);
    std::cout << label << ": " << hull.size() << " hull vertices:";
    for (const Point& p : hull) {
        std::cout << " (" << p.x << "," << p.y << ")";
    }
    std::cout << "   twice the area = " << twiceArea(hull) << '\n';
}

int main() {
    report("square with edge points",
           {{2, 2}, {4, 0}, {0, 2}, {1, 1}, {4, 4}, {2, 0}, {3, 1}, {0, 4}, {4, 2}, {2, 4}, {0, 0}, {1, 3}});
    report("triangle with edge points",
           {{3, 3}, {1, 1}, {6, 0}, {3, 0}, {2, 2}, {0, 0}, {4, 2}, {2, 1}});
    report("pentagon", {{1, 0}, {4, 1}, {5, 4}, {2, 6}, {0, 3}, {2, 3}, {3, 2}});
    return 0;
}

File: CSC310-Python/Module15-Computational-Geometry/debug02/convex_hull.py
"""Convex hull by Graham's scan (CLRS Section 33.3).

convex_hull(points) picks the lowest point (leftmost on ties) as the
anchor, sorts the remaining points by polar angle around the anchor
(nearer points first when angles tie), and then sweeps through them with
a stack, popping whenever the top of the stack would fail to make a left
turn.  The result is the hull's corner points in counter-clockwise order
starting at the anchor; a point that lies on a hull edge but is not a
corner is not part of the answer.  The main block runs three point sets
and prints twice the polygon area as a check.
"""

import math


def cross(o, a, b):
    """Cross product (a - o) x (b - o): positive for a left turn o->a->b."""
    return (a[0] - o[0]) * (b[1] - o[1]) - (a[1] - o[1]) * (b[0] - o[0])


def convex_hull(points):
    anchor = min(points, key=lambda p: (p[1], p[0]))
    rest = [p for p in points if p != anchor]
    rest.sort(key=lambda p: (math.atan2(p[1] - anchor[1], p[0] - anchor[0]),
                             (p[0] - anchor[0]) ** 2 + (p[1] - anchor[1]) ** 2))
    hull = [anchor]
    for p in rest:
        while len(hull) > 1 and cross(hull[-2], hull[-1], p) < 0:
            hull.pop()
        hull.append(p)
    return hull


def twice_area(polygon):
    total = 0
    for i in range(len(polygon)):
        x1, y1 = polygon[i]
        x2, y2 = polygon[(i + 1) % len(polygon)]
        total += x1 * y2 - x2 * y1
    return total


def report(title, points):
    hull = convex_hull(points)
    listing = " ".join(f"({x},{y})" for x, y in hull)
    print(f"{title}: {len(hull)} hull vertices: {listing}   "
          f"twice the area = {twice_area(hull)}")


def main():
    square = [(0, 0), (4, 0), (4, 4), (0, 4), (2, 0), (4, 2), (2, 4), (0, 2),
              (1, 1), (2, 2), (3, 1)]
    triangle = [(0, 0), (6, 0), (3, 3), (2, 0), (4, 0), (1, 1), (2, 2),
                (5, 1), (4, 2), (3, 1)]
    pentagon = [(1, 0), (4, 1), (5, 4), (2, 6), (0, 3), (2, 2), (3, 3),
                (1, 4), (3, 1)]
    report("square with edge points", square)
    report("triangle with edge points", triangle)
    report("pentagon", pentagon)


if __name__ == "__main__":
    main()
